src/lua/datetime.lua

```diff
@@ -420,6 +612,228 @@ local function datetime_tostring(self)
     return s
 end
 
+local function qtail(s)
+    return #s ~= 0 and (s .. ', ') or s
+end
+
+--[[
+    Convert to text interval values of different types
+
+    - depending on a values stored there generic interval
+      values may display in following format:
+        +12 secs
+        -23 minutes, 0 seconds
+        +12 hours, 23 minutes, 1 seconds
+        -7 days, -23 hours, -23 minutes, -1 seconds
+    - years will be displayed as
+        +10 years
+    - months will be displayed as:
+         +2 months
+]]
+local function interval_tostring(o)
+    check_interval(o, 'datetime.interval.tostring')
+    local s = ''
+    if o.year ~= 0 then
+        s = qtail(s) .. ('%+d years'):format(o.year)
+    end
+    if o.month ~= 0 then
+        s = qtail(s) .. ('%+d months'):format(o.month)
+    end
+    local sec, nsec = o.sec, o.nsec
+    if sec == 0 and nsec == 0 then
+        return #s > 0 and s or '0 seconds'
+    end
+    local abs_s = math_abs(sec)
+    if abs_s < 60 then
+        s = qtail(s) .. ('%+.9g seconds'):format(sec)
+    elseif abs_s < (60 * 60) then
+        s = qtail(s) .. ('%+d minutes, %.9g seconds'):
+                        format(o.min, math_fmod(sec, 60))
+    elseif abs_s < SECS_PER_DAY then
+        s = qtail(s) .. ('%+d hours, %d minutes, %.9g seconds'):
+                        format(o.hour, math_fmod(o.min, 60), math_fmod(sec, 60))
+    else
+        s = qtail(s) .. ('%+d days, %d hours, %d minutes, %.9g seconds'):
+                        format(o.day, math_fmod(o.hour, 24),
+                               math_fmod(o.min, 60), math_fmod(sec, 60))
+    end
+    if nsec > 0 then
+        s = qtail(s) .. ('%d nanoseconds'):format(nsec)
+    end
+    return s
+end
+
+local adjust_xlat = {
+    none = builtin.DT_LIMIT,
+    last = builtin.DT_SNAP,
+    excess = builtin.DT_EXCESS,
+}
+
+local function datetime_increment_by(self, direction, years, months,
```

Comment:
Something strange:
```lua
tarantool> dt.new{year=2000, month=3, day=1} + {year=1}
---
- 2001-03-02T00:00:00Z
...


```
Should result in `2001-03-01T00:00:00Z`, shouldn't it?

Reply:
After some further debugging I've found the root of a problem here - `(tnt_)?dt_add_years` simply use the day in year number for advancing year back or forward. And if such move spans over boundaries of leap year, then we could found ourselves on different month/day.

OTOH, `(tnt_)?dt_add_months` properly handle span to last day, or limits, or excesses. So as a work-around I always use `tnt_dt_add_months` even for year intervals operations.

Also below you see additional patch which is making `adjust` be part of 1st argument object - the way we agreed upon after discussion offline and the way we finalized in RFC since recently.

```diff
diff --git a/extra/exports b/extra/exports
index 88db07b91..15faabb12 100644
--- a/extra/exports
+++ b/extra/exports
@@ -420,7 +420,6 @@ tnt_datetime_strftime
 tnt_datetime_to_string
 tnt_default_cert_dir_paths
 tnt_default_cert_file_paths
-tnt_dt_add_years
 tnt_dt_add_months
 tnt_dt_days_in_month
 tnt_dt_dom
diff --git a/src/lua/datetime.lua b/src/lua/datetime.lua
index fbd250a4b..695fe1e9d 100644
--- a/src/lua/datetime.lua
+++ b/src/lua/datetime.lua
@@ -63,7 +63,6 @@ typedef enum {
     DT_SNAP
 } dt_adjust_t;
 
-dt_t   tnt_dt_add_years    (dt_t dt, int delta, dt_adjust_t adjust);
 dt_t   tnt_dt_add_months   (dt_t dt, int delta, dt_adjust_t adjust);
 
 /* dt_parse_iso.h definitions */
@@ -154,6 +153,7 @@ ffi.cdef [[
         int nsec;
         int month;
         int year;
+        dt_adjust_t adjust;
     };
 ]]
 local interval_t = ffi.typeof('struct datetime_interval')
@@ -319,12 +319,19 @@ local function normalize_nsec(secs, nsec)
     return secs, nsec
 end
 
+local adjust_xlat = {
+    none = builtin.DT_LIMIT,
+    last = builtin.DT_SNAP,
+    excess = builtin.DT_EXCESS,
+}
+
 local function interval_decouple_args(obj)
     if is_interval(obj) then
-        return obj.year, obj.month, obj.sec, obj.nsec
+        return obj.year, obj.month, obj.sec, obj.nsec, obj.adjust
     end
     local year = checked_max_value(obj.year, MAX_YEAR_RANGE, 'year')
     local month = checked_max_value(obj.month, MAX_MONTH_RANGE, 'month')
+    local adjust = adjust_xlat[obj.adjust] or builtin.DT_LIMIT
 
     local secs = 0
     secs = secs + (7 * SECS_PER_DAY) *
@@ -355,7 +362,7 @@ local function interval_decouple_args(obj)
     nsec = (msec or 0) * 1e6 + (usec or 0) * 1e3 + (nsec or 0)
     secs, nsec = normalize_nsec(secs, nsec)
 
-    return year, month, secs, nsec
+    return year, month, secs, nsec, adjust
 end
 
 local function interval_new(obj)
@@ -364,12 +371,13 @@ local function interval_new(obj)
         return ival
     end
     check_table(obj, 'interval.new()')
-    ival.year, ival.month, ival.sec, ival.nsec = interval_decouple_args(obj)
+    ival.year, ival.month, ival.sec, ival.nsec, ival.adjust =
+        interval_decouple_args(obj)
     return ival
 end
 
-local function interval_init(year, month, sec, nsec)
-    return ffi.new(interval_t, sec, nsec, month, year)
+local function interval_init(year, month, sec, nsec, adjust)
+    return ffi.new(interval_t, sec, nsec, month, year, adjust)
 end
 
 local function nyi(msg)
@@ -663,12 +671,6 @@ local function interval_tostring(o)
     return s
 end
 
-local adjust_xlat = {
-    none = builtin.DT_LIMIT,
-    last = builtin.DT_SNAP,
-    excess = builtin.DT_EXCESS,
-}
-
 local function datetime_increment_by(self, direction, years, months,
                                      seconds, nanoseconds, adjust)
     -- operations with intervals should be done using local human dates
@@ -678,19 +680,15 @@ local function datetime_increment_by(self, direction, years, months,
     local offset = self.tzoffset
 
     local is_ym_updated = false
-    if adjust ~= nil then
-        local operation = direction >= 0 and 'addition' or 'subtraction'
-        check_table(adjust, operation)
-        adjust = adjust_xlat[adjust.adjust] or builtin.DT_LIMIT
-    else
-        adjust = builtin.DT_LIMIT
-    end
+    adjust = adjust or builtin.DT_LIMIT
 
     if years ~= 0 then
         check_range(years, MIN_DATE_YEAR, MAX_DATE_YEAR, 'years')
         local approx_dt = dt + direction * years * AVERAGE_DAYS_YEAR
         check_dt(approx_dt, direction, years, 'years')
-        dt = builtin.tnt_dt_add_years(dt, direction * years, adjust)
+        -- tnt_dt_add_years() not handle properly DT_SNAP or DT_LIMIT mode
+        -- so use tnt_dt_add_months() as a work-around
+        dt = builtin.tnt_dt_add_months(dt, direction * years * 12, adjust)
         is_ym_updated = true
     end
     if months ~= 0 then
@@ -746,13 +744,13 @@ local function datetime_interval_sub(lhs, rhs)
     check_date_interval_table(rhs, "operator -")
     local left_interval = is_table(lhs) or ffi.typeof(lhs) == interval_t
     local right_interval = is_table(rhs) or ffi.typeof(rhs) == interval_t
-    local year, month, secs, nsec
+    local year, month, secs, nsec, adjust
 
     -- left is date, right is interval
     if not left_interval and right_interval then
-        year, month, secs, nsec = interval_decouple_args(rhs)
+        year, month, secs, nsec, adjust = interval_decouple_args(rhs)
         return datetime_increment_by(datetime_new_copy(lhs), -1, year, month,
-                                     secs, nsec)
+                                     secs, nsec, adjust)
     -- left is date, right is date
     elseif not left_interval and not right_interval then
         local obj = interval_new()
@@ -761,8 +759,8 @@ local function datetime_interval_sub(lhs, rhs)
         return obj
     -- both left and right are intervals
     elseif left_interval and right_interval then
-        year, month, secs, nsec = interval_decouple_args(lhs)
-        local obj = interval_init(year, month, secs, nsec)
+        year, month, secs, nsec, adjust = interval_decouple_args(lhs)
+        local obj = interval_init(year, month, secs, nsec, adjust)
         year, month, secs, nsec = interval_decouple_args(rhs)
         obj.year = obj.year - year
         obj.month = obj.month - month
@@ -789,17 +787,17 @@ local function datetime_interval_add(lhs, rhs)
     check_interval_table(rhs, "operator +")
     local left_interval = is_table(lhs) or ffi.typeof(lhs) == interval_t
     local right_interval = is_table(rhs) or ffi.typeof(rhs) == interval_t
-    local year, month, secs, nsec
+    local year, month, secs, nsec, adjust
 
     -- left is date, right is interval
     if not left_interval and right_interval then
         local obj = datetime_new_copy(lhs)
-        year, month, secs, nsec = interval_decouple_args(rhs)
-        return datetime_increment_by(obj, 1, year, month, secs, nsec)
+        year, month, secs, nsec, adjust = interval_decouple_args(rhs)
+        return datetime_increment_by(obj, 1, year, month, secs, nsec, adjust)
     -- both left and right are intervals
     elseif left_interval and right_interval then
-        year, month, secs, nsec = interval_decouple_args(lhs)
-        local obj = interval_init(year, month, secs, nsec)
+        year, month, secs, nsec, adjust = interval_decouple_args(lhs)
+        local obj = interval_init(year, month, secs, nsec, adjust)
         year, month, secs, nsec = interval_decouple_args(rhs)
         obj.year = obj.year + year
         obj.month = obj.month + month
@@ -846,13 +844,14 @@ end
 
 -- addition or subtraction from date/time of a given interval
 -- described via table direction should be +1 or -1
-local function datetime_shift(self, o, direction, adjust)
+local function datetime_shift(self, o, direction)
     assert(direction == -1 or direction == 1)
     local title = direction > 0 and "datetime.add" or "datetime.sub"
     check_interval_table(o, title)
 
-    local year, month, secs, nsec = interval_decouple_args(o)
-    return datetime_increment_by(self, direction, year, month, secs, nsec, adjust)
+    local year, month, secs, nsec, adjust = interval_decouple_args(o)
+    return datetime_increment_by(self, direction, year, month, secs, nsec,
+                                 adjust)
 end
 
 --[[
@@ -1090,12 +1089,8 @@ local datetime_index_functions = {
     format = datetime_format,
     totable = datetime_totable,
     set = datetime_set,
-    add = function(self, obj, adjust)
-            return datetime_shift(self, obj, 1, adjust)
-          end,
-    sub = function(self, obj, adjust)
-            return datetime_shift(self, obj, -1, adjust)
-          end,
+    add = function(self, obj) return datetime_shift(self, obj, 1) end,
+    sub = function(self, obj) return datetime_shift(self, obj, -1) end,
 }
 
 local function datetime_index(self, key)
```

<details>
  <summary>Test changed collapsed below...</summary>
  
```diff
diff --git a/test/app-tap/datetime.test.lua b/test/app-tap/datetime.test.lua
index 826a02595..bfbdfefc5 100755
--- a/test/app-tap/datetime.test.lua
+++ b/test/app-tap/datetime.test.lua
@@ -664,13 +664,13 @@ test:test("Time interval operations", function(test)
         {{month = 1}, '1972-03-29T00:00:00Z', '1972-01-29T00:00:00Z'},
         {{month = 1}, '1972-04-29T00:00:00Z', '1971-12-29T00:00:00Z'},
         {{month = 5}, '1972-09-29T00:00:00Z', '1971-07-29T00:00:00Z'},
-        {{year = 1, month = 2}, '1973-11-30T00:00:00Z', '1970-05-29T00:00:00Z'},
-        {{year = 2, month = 3}, '1976-02-29T00:00:00Z', '1968-02-28T00:00:00Z'},
-        {{month = 1}, '1976-03-29T00:00:00Z', '1968-01-28T00:00:00Z'},
-        {{month = 1}, '1976-04-29T00:00:00Z', '1967-12-28T00:00:00Z'},
-        {{month = 5}, '1976-09-29T00:00:00Z', '1967-07-28T00:00:00Z'},
-        {{year = -1}, '1975-09-30T00:00:00Z', '1968-07-27T00:00:00Z'},
-        {{year = -1}, '1974-09-30T00:00:00Z', '1969-07-28T00:00:00Z'},
+        {{year = 1, month = 2}, '1973-11-29T00:00:00Z', '1970-05-29T00:00:00Z'},
+        {{year = 2, month = 3}, '1976-02-29T00:00:00Z', '1968-02-29T00:00:00Z'},
+        {{month = 1}, '1976-03-29T00:00:00Z', '1968-01-29T00:00:00Z'},
+        {{month = 1}, '1976-04-29T00:00:00Z', '1967-12-29T00:00:00Z'},
+        {{month = 5}, '1976-09-29T00:00:00Z', '1967-07-29T00:00:00Z'},
+        {{year = -1}, '1975-09-29T00:00:00Z', '1968-07-29T00:00:00Z'},
+        {{year = -1}, '1974-09-29T00:00:00Z', '1969-07-29T00:00:00Z'},
     }
 
     local tadd_t = date.new(base_leap_date) -- :add{table}
@@ -815,20 +815,20 @@ test:test("Time interval operations - different adjustments", function(test)
                        '1972-04-29T00:00:00Z'}},
         {{month = 5}, {'1972-09-30T00:00:00Z', '1972-09-29T00:00:00Z',
                        '1972-09-29T00:00:00Z'}},
-        {{year = 1, month = 2}, {'1973-12-01T00:00:00Z', '1973-11-30T00:00:00Z',
-                                 '1973-11-30T00:00:00Z'}},
-        {{year = 2, month = 3}, {'1976-03-01T00:00:00Z', '1976-02-29T00:00:00Z',
-                                 '1976-03-01T00:00:00Z'}},
-        {{month = 1}, {'1976-04-01T00:00:00Z', '1976-03-29T00:00:00Z',
-                       '1976-04-01T00:00:00Z'}},
-        {{month = 1}, {'1976-05-01T00:00:00Z', '1976-04-29T00:00:00Z',
-                       '1976-05-01T00:00:00Z'}},
-        {{month = 5}, {'1976-10-01T00:00:00Z', '1976-09-29T00:00:00Z',
-                       '1976-10-01T00:00:00Z'}},
-        {{year = -1}, {'1975-10-02T00:00:00Z', '1975-09-30T00:00:00Z',
-                       '1975-10-02T00:00:00Z'}},
-        {{year = -1}, {'1974-10-02T00:00:00Z', '1974-09-30T00:00:00Z',
-                       '1974-10-02T00:00:00Z'}},
+        {{year = 1, month = 2}, {'1973-11-30T00:00:00Z', '1973-11-29T00:00:00Z',
+                                 '1973-11-29T00:00:00Z'}},
+        {{year = 2, month = 3}, {'1976-02-29T00:00:00Z', '1976-02-29T00:00:00Z',
+                                 '1976-02-29T00:00:00Z'}},
+        {{month = 1}, {'1976-03-31T00:00:00Z', '1976-03-29T00:00:00Z',
+                       '1976-03-29T00:00:00Z'}},
+        {{month = 1}, {'1976-04-30T00:00:00Z', '1976-04-29T00:00:00Z',
+                       '1976-04-29T00:00:00Z'}},
+        {{month = 5}, {'1976-09-30T00:00:00Z', '1976-09-29T00:00:00Z',
+                       '1976-09-29T00:00:00Z'}},
+        {{year = -1}, {'1975-09-30T00:00:00Z', '1975-09-29T00:00:00Z',
+                       '1975-09-29T00:00:00Z'}},
+        {{year = -1}, {'1974-09-30T00:00:00Z', '1974-09-29T00:00:00Z',
+                       '1974-09-29T00:00:00Z'}},
     }
 
     local tadd = {}
@@ -840,10 +840,12 @@ test:test("Time interval operations - different adjustments", function(test)
     for _, row in pairs(leap_ivals_check_add) do
         local delta, sadds = unpack(row)
         for i=1,3 do
+            local tsbefore = tostring(tadd[i])
             local adjust = adj_modes[i]
             local sadd = sadds[i]
-            test:is(tostring(tadd[i]:add(delta, {adjust=adjust})), sadd,
-                    ('add delta to %s with adjust %s'):format(sadd, adjust))
+            delta.adjust = adjust
+            test:is(tostring(tadd[i]:add(delta)), sadd,
+                    ('add delta to %s with adjust %s'):format(tsbefore, adjust))
 
         end
     end
@@ -857,18 +859,18 @@ test:test("Time interval operations - different adjustments", function(test)
                        '1971-07-29T00:00:00Z'}},
         {{year = 1, month = 2}, {'1970-05-31T00:00:00Z', '1970-05-29T00:00:00Z',
                        '1970-05-29T00:00:00Z'}},
-        {{year = 2, month = 3}, {'1968-02-29T00:00:00Z', '1968-02-28T00:00:00Z',
-                       '1968-02-28T00:00:00Z'}},
-        {{month = 1}, {'1968-01-31T00:00:00Z', '1968-01-28T00:00:00Z',
-                       '1968-01-28T00:00:00Z'}},
-        {{month = 1}, {'1967-12-31T00:00:00Z', '1967-12-28T00:00:00Z',
-                       '1967-12-28T00:00:00Z'}},
-        {{month = 5}, {'1967-07-31T00:00:00Z', '1967-07-28T00:00:00Z',
-                       '1967-07-28T00:00:00Z'}},
-        {{year = -1}, {'1968-07-30T00:00:00Z', '1968-07-27T00:00:00Z',
-                       '1968-07-27T00:00:00Z'}},
-        {{year = -1}, {'1969-07-31T00:00:00Z', '1969-07-28T00:00:00Z',
-                       '1969-07-28T00:00:00Z'}},
+        {{year = 2, month = 3}, {'1968-02-29T00:00:00Z', '1968-02-29T00:00:00Z',
+                       '1968-02-29T00:00:00Z'}},
+        {{month = 1}, {'1968-01-31T00:00:00Z', '1968-01-29T00:00:00Z',
+                       '1968-01-29T00:00:00Z'}},
+        {{month = 1}, {'1967-12-31T00:00:00Z', '1967-12-29T00:00:00Z',
+                       '1967-12-29T00:00:00Z'}},
+        {{month = 5}, {'1967-07-31T00:00:00Z', '1967-07-29T00:00:00Z',
+                       '1967-07-29T00:00:00Z'}},
+        {{year = -1}, {'1968-07-31T00:00:00Z', '1968-07-29T00:00:00Z',
+                       '1968-07-29T00:00:00Z'}},
+        {{year = -1}, {'1969-07-31T00:00:00Z', '1969-07-29T00:00:00Z',
+                       '1969-07-29T00:00:00Z'}},
     }
 
     local tsub = {}
@@ -880,10 +882,12 @@ test:test("Time interval operations - different adjustments", function(test)
     for _, row in pairs(leap_ivals_check_sub) do
         local delta, ssubs = unpack(row)
         for i=1,3 do
+            local tsbefore = tostring(tsub[i])
             local adjust = adj_modes[i]
             local ssub = ssubs[i]
-            test:is(tostring(tsub[i]:sub(delta, {adjust=adjust})), ssub,
-                    ('sub delta to %s with adjust %s'):format(ssub, adjust))
+            delta.adjust = adjust
+            test:is(tostring(tsub[i]:sub(delta)), ssub,
+                    ('sub delta to %s with adjust %s'):format(tsbefore, adjust))
         end
     end
 end)
@@ -941,7 +945,7 @@ end)
 
 test:test("Months intervals with last days", function(test)
     test:plan(122)
-    local month1 = date.interval.new{month = 1}
+    local month1 = date.interval.new{month = 1, adjust = 'last'}
     local base_day = date.new{year = 1998, month = 12, day = 31}
     local current = base_day
     test:is(current, date.new{year = 1998, month = 12, day = -1}, "day = -1")

```
</details>

Reply:
@sergepetrenko:
> I don't see the corresponding change in the RFC. Is it still here?

Yes, it was modified in the chapter https://hackmd.io/hJU7rbnGQv6CSJCJl7KUiA?view#%D0%9F%D0%B0%D1%80%D0%B0%D0%BC%D0%B5%D1%82%D1%80-adjust

i.e. instead of additional object argument passed separately, `adjust` passed at the same object where interval described.

Reply:
@sergepetrenko:
> BTW, I was going to ask which adjust parameter is assigned to an interval coming from 2 dates subtraction, 
> but I noticed such a subtraction always results in an interval with days, but not months or years. 
> So adjust is not needed in that case. Is this intended?

Not so much intended, but is the only possible (of simplest) way(s). When we have both dates to the left and to the right we simply create interval of difference `epoch` and `nsec` parts. 

```
tarantool> a = date.new{year = 2001}
---
...

tarantool> b = date.new{year = 2002}
---
...

tarantool> b - a
---
- +365 days, 0 hours, 0 minutes, 0 seconds
...

tarantool> I = b - a
---
...

tarantool> ffi.typeof(I)
---
- ctype<struct datetime_interval>
...

tarantool> I.sec
---
- 31536000
...

tarantool> I.nsec
---
- 0
...

tarantool> I.adjust
---
- 'cdata<enum 997>: 1'
...

```

(where 1 is `DT_LIMIT`)

```lua
    -- left is date, right is date
    elseif not left_interval and not right_interval then
        local obj = interval_new()
        obj.sec, obj.nsec = normalize_nsec(lhs.epoch - rhs.epoch,
                                           lhs.nsec - rhs.nsec)
        return obj
```